### ISSUE TITLE
Generalize indexing primitives to work with type Integer

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/cryptol#502]
Note: This ticket was previously a part of #347.

It would be nice to generalize some Cryptol primitives so that they can work with either `[n]` or `Integer`:
- list indexes for `@`, `!`, `@@`, `!!`, and `update`
- shift and rotation amounts for `<<`, `>>`, `<<<`, and `>>>`
- list enumerations `[x...]`, `[x,y...]`, `[a..b]`, and `[a,b..c]`
  We could introduce a new class (maybe `Index a`) that would have instances for `[n]` and `Integer`, and generalize the types of the above primitives to use this class constraint, for example:

```
primitive (<<<) : {n, a, b} (fin n, Index b) => [n]a -> b -> [n]a
primitive (@) : {n, a, b} (Index b) => [n]a -> b -> a
```

See also #482 for generalizing `demote` to include type `Integer`.

**Comment by brianhuffman** [comment]
There is a design decision we have to make about how to handle negative indexes of type `Integer`. With `(@)`, `(!)`, and `update`, a negative index should produce an "invalid sequence index" error, just like any other out-of-bounds index. But for shifts and rotates, two choices make sense:
1. Shift/rotate by a negative amount is an error.
2. Shift/rotate by a negative amount is a shift/rotate in the opposite direction.

Which choice should we make?

**Comment by weaversa** [comment]
2.

**Comment by brianhuffman** [comment]
As of 9fcb48116183dce67abe9838321acd66a8216630, the various list enumerations now have generalized types:

- `[a..b]` and `[a,b..c]` are generalized to class `Literal`
- `[x...]` and `[x,y...]` are generalized to class `Arith`
- `[x..]` and `[x,y..]` are defined to go until the numbers would overflow mod 2^^n, so remain specialized to bitvectors

**Comment by kiniry** [comment]
Has the documentation been updated to correspond to this change? Let’s not let all these new features get ahead of us before we update the book and the cheat sheet.

**Comment by brianhuffman** [comment]
The "Programming Cryptol" book doesn't even mention the types of the list enumeration operators. I see that `CryptolPrims.pdf` is out of date in a lot of ways, so I'm updating that now.

**Comment by brianhuffman** [comment]
A couple of months ago I started implementing this feature on the `index-class` branch (see commit 2504bbe8d7afeecc03b6eb17027affcaafe70a2a).

One thing that will be tricky to implement is indexing an infinite list `[inf]a` with an `Integer` index: For concrete values it's easy, but doing the right thing for symbolic evaluation will require a lot of work.

**Comment by weaversa** [comment]
Once `@@` works with `Integer`, it would be great to have the type system default the indices to `Integer` rather than some `[a]`, especially for constant lists. Then maybe some of the warnings could go away. For example:

```
Cryptol> (zero : [10]) @@ [1..5]
[warning] at <interactive>:1:1--1:24:
  Defaulting type argument 'ix' of '(@@)' to 3
0x00
```
That warning always really bugs me. :-)

**Comment by robdockins** [comment]
Fixed via #724.  Indexing and shifting operations are now genealized over `Integral` which includes both bitvectors and `Integer`.
